### ISSUE TITLE
Add workaround for balances in last 15 mins

### ISSUE THREAD

**Issue by Nana-EC** [hashgraph/hedera-json-rpc-relay#694]
### Problem

Currently the Mirror Node doesn't not provide accurate balance to the given timestamp for values in the last 15 mins.
This provides an unexpected error for tools

### Solution

Explore a clean way to get balance for account within last 15 mins. 
Potentially when `eth_getBalance` for a block from the past 15 mins is called

- get the last balance and timestamp for an account from api/v1/balances based on a block timestamp
- get all the transactions for the account between the requested block timestamp and the last balance file using `api/v1/transactions?account.id={accountId}&timestamp=lte:{blockTimestamp}&timestamp=gt:{lastBalanceRecordTimestamp}`. Note this will require looping through 
- Sum up the credits and debits involving the account to get the last balance
- On the relay cache this lastBalance and timestamp with a `ttl` up to the last balance file
- Consult the balance cache on future calls. if present call transaction for credit/debits since that time

**Comment by Nana-EC** [comment]
Hey @mshakeg 
Thanks for the suggestion.

Can you clarify how the `IHUtil` flow would be different from the [IERC20](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/examples/erc-20/ERC20Contract.sol) or [IERC721](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/examples/erc-721/ERC721Contract.sol) interfaces already implemented?
Using that you can get an HTS tokens balance and other items per those interfaces?

In this case however it's more of an issue of the EOA's balance in terms of HBAR than the tokens

Also I think potentially part of the issue you're addressing here is the current requirement to have a wrapper contract that calls precompile HTS functions right? Would your issue be resolved if you could make direct calls to [IHederaTokenService](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol) as if it were a deployed contract?

**Comment by mshakeg** [comment]
@Nana-EC one of the major benefits to developers using the `IHUtil` flow would be in reducing contract bytecode size as instead of having to import and reference the [SafeHts](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/safe-hts-precompile/SafeHTS.sol) library or any of the other precompile contracts explicitly an `HUtil` contract would be deployed and developers can reduce their contract bytecode by referencing that `HUtil` contract via the `IHUtil` interface as shown in `DemoContract`. Also having a smaller code base allows for higher optimizer runs with less worry of exceeding the bytecode size limit.

Also instead of each developer having to deploy and pay for renewal fees of the `HUtil` contract, Hedera could manage and update a single `HUtil` contract deployed to a static account/contract id that all developers can reference in their contracts, and EOAs can interact with as explained below.

Another advantage of the `HUtil` contract is that users/EOAs can interact with HTS, HCS and any other services integrated with the HSCS using existing Ethereum-compatible wallets(provided they can connect to a custom relay). For example, a Metamask user can associate their account with an HTS token by sending a transaction to `HUtil.associateSender` for example, etc for other functions(transfers, dissociations, ...) exposed by `HUtil`. This would be helpful as existing Metamask users don't have to setup an account in Hashpack, etc to have a functional account on Hedera.

In summary, it's useful to both developers and end-users.

More specific to this issue the `HUtil.balanceOf` function could be generalised to get the HBAR balance or expose another function say `nativeBalance` to get the native(HBAR) balance.
```solidity

// Method 1:
function balanceOf(address token, address account) public view returns (uint) {

    if (token == address(0)) { // interpret address(0) as native token
        return address(account).balance;
    } else {
        return IERC20(token).balanceOf(account);
    }

}

// Method 2:
function nativeBalance(address account) public view returns (uint) {

    return address(account).balance;

}
```

**Comment by mshakeg** [comment]
Looked a bit more into the existing functionality and learnt that the [supported ERC token standards](https://docs.hedera.com/guides/core-concepts/smart-contracts/supported-erc-token-standards) aren't only accessible internally to other contracts but also to EOAs. So that would make EOAs/users using Ethereum compatible wallets such as Metamask able to import their HTS tokens like ERC tokens, in addition, devs don't have to use any of the HTS precompiles for all the functionality that is exposed by the ERC token interfaces.

However, for functionality that is not exposed by the ERC token standards that are specific to Hedera such as token associations and dissociations perhaps the ERC interfaces could be extended to support that Hedera specific functionality and completely eliminate the need to use any of the HTS precompiles.

Also if ERC calls&txs from EOAs are currently being routed to the EVM it might be a good idea to reduce the load on the EVM and tx price(since HTS interactions from the HSCS are slightly more expensive than direct HTS txs) by decoding the eth transaction and having the HTS layer handle it directly.

**Comment by Nana-EC** [comment]
Hi @mshakeg, 
Let's move this to a separate issue/discussion to make sure we can explore your suggestion as I think it's a little different from the ticket description.

Depending on the suggestion I think you may require a HIP discussion and eventually a HIP.

Restating my question here but

> Would your issue be resolved if you could make direct calls to [IHederaTokenService](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol) as if it were a deployed contract? 
This would support all the noted methods without having to create a wrapper contract s you noted
